core/base/octree/Octree.h

```diff
@@ -0,0 +1,215 @@
+/// \ingroup base
+/// \class Octree
+/// \author Guoxi Liu (guoxil@g.clemson.edu)
+/// \date May 2021.
+///
+/// \brief Implementation of the point region (PR) octree.
+///
+/// \b Related \b publications \n
+/// "The PR-star octree: A spatio-topological data structure for tetrahedral
+/// meshes." Kenneth Weiss, Leila Floriani, Riccardo Fellegara, and Marcelo
+/// Velloso In Proceedings of the 19th ACM SIGSPATIAL International Conference
+/// on Advances in Geographic Information Systems, 2011.
+///
+/// \sa ttk::CompactTriangulationPreconditioning
+
+#pragma once
+
+#include <Triangulation.h>
+#include <assert.h>
+#include <iostream>
+#include <stack>
+#include <unordered_map>
+#include <vector>
+
+using namespace ttk;
+using namespace std;
+
+class OctreeNode {
+public:
+  OctreeNode() {
+    OctreeNode(0);
+  }
+  OctreeNode(uint32_t location) {
+    locCode = location;
+    childExists = 0;
+  }
+
+  ~OctreeNode() {
+  }
+
+protected:
+  uint32_t locCode;
+  uint8_t childExists;
+  vector<SimplexId> vertexIds;
+  vector<SimplexId> cellIds;
+
+  friend class Octree;
+};
+
+class Octree : public virtual Debug {
+public:
+  Octree(const AbstractTriangulation *t);
+  Octree(const AbstractTriangulation *t, const int k);
+  ~Octree();
+
+  bool empty();
+
+  size_t getNodeTreeDepth(const OctreeNode *node);
+
+  OctreeNode *getParentNode(OctreeNode *node);
+
+  void visitAll(const OctreeNode *node);
+
+  int verifyTree(SimplexId &vertexNum);
+
+  int insertVertex(SimplexId &vertexId);
+
+  int insertCell(SimplexId &cellId);
+
+  void reindex(vector<SimplexId> &vertices,
+               vector<SimplexId> &nodes,
+               vector<SimplexId> &cells);
+
+private:
+  const AbstractTriangulation *triangulation_;
+  unordered_map<uint32_t, OctreeNode> allNodes;
+  int capacity;
+  float center[3];
+  float size[3];
```

Comment:
Maybe use `std::array<float, 3>` instead of C-array? `std::array` are zero-initialized and have additional methods (`begin()`, `end()`, ...) that make them compatible with `std::vector`. To get a C-array like pointer, use the `data()` method.